src/scheduler.cpp

```diff
@@ -118,15 +114,15 @@ void CScheduler::MockForward(std::chrono::seconds delta_seconds)
     newTaskScheduled.notify_one();
 }
 
-static void Repeat(CScheduler* s, CScheduler::Function f, int64_t deltaMilliSeconds)
+static void Repeat(CScheduler& s, CScheduler::Function f, std::chrono::milliseconds delta)
```

Comment:
Nit: `const std::chrono::milliseconds&` to avoid copy (here and below).

Reply:
Nit on nit:

I'm afraid that might be a pessimisation :)

From a performance perspective the typical rule-of-thumb recommendation is to pass "in" parameters by value (instead of reference to const) if the parameter has a size up to 16 bytes ("cheaply-copied" types).

Thus in this case the rule-of-thumb recommendation would be to pass by value instead of reference to const.

(Disclaimer: As always with optimisations -- if it is worth trying to optimise then it is worth trying to quantify the impact of the optimisation by measuring! :))

From the C++ Core Guidelines:

> [F.16: For “in” parameters, pass cheaply-copied types by value and others by reference to const](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f16-for-in-parameters-pass-cheaply-copied-types-by-value-and-others-by-reference-to-const)
>
> Reason: Both let the caller know that a function will not modify the argument, and both allow initialization by rvalues.
>
> What is “cheap to copy” depends on the machine architecture, but two or three words (doubles, pointers, references) are usually best passed by value. When copying is cheap, nothing beats the simplicity and safety of copying, and for small objects (up to two or three words) it is also faster than passing by reference because it does not require an extra indirection to access from the function.

![](https://isocpp.github.io/CppCoreGuidelines/param-passing-normal.png)

Reply:
It feels unstable to me to depend on architecture and/or implementation rather than just go for const x& when you can, but it's a nit; I'll live. :)

Reply:
`std::chrono::milliseconds` is just a 64-bit int (well, 45-bit or greater int), so passing it as `const &` means you're passing a 64-bit pointer instead, and have to deref it later.

Reply:
Leaving as is for now, to not invalidate review

Reply:
The other feedback aside, I prefer the copy here over a reference because other arguments (like the function) *must* be copied anyway.

Reply:
Yep, copy seems sensible after all.